src/core/algorithms/dc/FastADC/fastadc.cpp

```diff
@@ -0,0 +1,147 @@
+    RegisterOption(Option{&shard_length_, kShardLength, kDShardLength, 350U});
+    RegisterOption(Option{&allow_cross_columns_, kAllowCrossColumns, kDAllowCrossColumns, true});
+    RegisterOption(Option{&minimum_shared_value_, kMinimumSharedValue, kDMinimumSharedValue, 0.3});
+    RegisterOption(
+            Option{&comparable_threshold_, kComparableThreshold, kDComparableThreshold, 0.1});
+    RegisterOption(Option{&evidence_threshold_, kEvidenceThreshold, kDEvidenceThreshold, 0.01});
+}
+
+void FastADC::MakeExecuteOptsAvailable() {
+    using namespace config::names;
+
+    MakeOptionsAvailable({kShardLength, kAllowCrossColumns, kMinimumSharedValue,
+                          kComparableThreshold, kEvidenceThreshold});
+}
+
+void FastADC::LoadDataInternal() {
+    typed_relation_ = model::ColumnLayoutTypedRelationData::CreateFrom(
+            *input_table_, true, true);  // kMixed type will be treated as a string type
+
+    if (typed_relation_->GetColumnData().empty()) {
+        throw std::runtime_error("Got an empty dataset: DC mining is meaningless.");
+    }
+}
+
+void FastADC::SetLimits() {
+    unsigned all_rows_num = typed_relation_->GetNumRows();
+
+    if (shard_length_ > all_rows_num) {
+        throw std::invalid_argument(
+                "'shard_length' (" + std::to_string(shard_length_) +
+                ") must be less or equal to the number of rows in the table (total "
+                "rows: " +
+                std::to_string(all_rows_num) + ")");
+    }
+    if (shard_length_ == 0) shard_length_ = all_rows_num;
+}
+
+void FastADC::CheckTypes() {
+    model::ColumnIndex columns_num = typed_relation_->GetNumColumns();
+    unsigned rows_num = typed_relation_->GetNumRows();
+
+    for (model::ColumnIndex column_index = 0; column_index < columns_num; column_index++) {
+        model::TypedColumnData const& column = typed_relation_->GetColumnData(column_index);
+        model::TypeId type_id = column.GetTypeId();
+
+        if (type_id == +model::TypeId::kMixed) {
+            LOG(WARNING) << "Column with index \"" + std::to_string(column_index) +
+                                    "\" contains values of different types. Those values will be "
+                                    "treated as strings.";
+        } else if (!column.IsNumeric() && type_id != +model::TypeId::kString) {
+            throw std::invalid_argument(
+                    "Column with index \"" + std::to_string(column_index) +
+                    "\" is of unsupported type. Only numeric and string types are supported.");
+        }
+
+        for (std::size_t row_index = 0; row_index < rows_num; row_index++) {
+            if (column.IsNull(row_index)) {
+                throw std::runtime_error("Some of the value coordinates are nulls.");
+            }
+            if (column.IsEmpty(row_index)) {
+                throw std::runtime_error("Some of the value coordinates are empty.");
+            }
```

Comment:
If it's not important you may combine with `column.IsNullOrEmpty()`
```suggestion
            if (column.IsNullOrEmpty(row_index)) {
                throw std::runtime_error("Some of the value coordinates are null or empty.");
            }
```